### ISSUE TITLE
[Bug] Possible `NameError` within constructed decorator

### ISSUE THREAD

**Issue by langfield** [beartype/beartype#158]
See below. I'm actually not sure I'm completely sure whether this is a @beartype thing or something specific to my code. But it is confusing and appears to break within a `<@beartype ...>` thing. :woman_shrugging: 

Relevant function:

```python

@beartype
def rmtree2(path: str) -> None:
    """On windows, rmtree fails for readonly dirs."""

    @beartype
    def handle_remove_readonly(
        func: Callable[..., Any],
        path: str,
        exc: Tuple[Type[OSError], OSError, TracebackType],
    ) -> None:
        excvalue = exc[1]
        if func in (os.rmdir, os.remove, os.unlink) and excvalue.errno == errno.EACCES:
            for p in (path, os.path.dirname(path)):
                os.chmod(p, os.stat(p).st_mode | stat.S_IWUSR)
            func(path)
        else:
            raise

    shutil.rmtree(path, ignore_errors=False, onerror=handle_remove_readonly)
```

```python
2022-08-13T14:34:32.8000113Z _______________________ test_computes_and_stores_md5sum _______________________
2022-08-13T14:34:32.8000438Z 
2022-08-13T14:34:32.8001352Z path = 'C:\\Users\\runneradmin\\AppData\\Local\\Temp\\tmpqv92rnle\\ki-local-199216c39eeabe23a1da016a99ffd3e2\\.git\\objects\\0a'
2022-08-13T14:34:32.8001967Z onerror = <function rmtree2.<locals>.handle_remove_readonly at 0x0000014F4AEBCA60>
2022-08-13T14:34:32.8002286Z 
2022-08-13T14:34:32.8002509Z     def _rmtree_unsafe(path, onerror):
2022-08-13T14:34:32.8002828Z         try:
2022-08-13T14:34:32.8003178Z             with os.scandir(path) as scandir_it:
2022-08-13T14:34:32.8003558Z                 entries = list(scandir_it)
2022-08-13T14:34:32.8003882Z         except OSError:
2022-08-13T14:34:32.8004273Z             onerror(os.scandir, path, sys.exc_info())
2022-08-13T14:34:32.8004633Z             entries = []
2022-08-13T14:34:32.8004965Z         for entry in entries:
2022-08-13T14:34:32.8005293Z             fullname = entry.path
2022-08-13T14:34:32.8005642Z             if _rmtree_isdir(entry):
2022-08-13T14:34:32.8005972Z                 try:
2022-08-13T14:34:32.8006478Z                     if entry.is_symlink():
2022-08-13T14:34:32.8006876Z                         # This can only happen if someone replaces
2022-08-13T14:34:32.8007359Z                         # a directory with a symlink after the call to
2022-08-13T14:34:32.8007675Z                         # os.scandir or entry.is_dir above.
2022-08-13T14:34:32.8007976Z                         raise OSError("Cannot call rmtree on a symbolic link")
2022-08-13T14:34:32.8008258Z                 except OSError:
2022-08-13T14:34:32.8010699Z                     onerror(os.path.islink, fullname, sys.exc_info())
2022-08-13T14:34:32.8010986Z                     continue
2022-08-13T14:34:32.8011258Z                 _rmtree_unsafe(fullname, onerror)
2022-08-13T14:34:32.8011495Z             else:
2022-08-13T14:34:32.8011703Z                 try:
2022-08-13T14:34:32.8012396Z >                   os.unlink(fullname)
2022-08-13T14:34:32.8013822Z E                   PermissionError: [WinError 5] Access is denied: 'C:\\Users\\runneradmin\\AppData\\Local\\Temp\\tmpqv92rnle\\ki-local-199216c39eeabe23a1da016a99ffd3e2\\.git\\objects\\0a\\04e0bdda860c1231ad35cef1f3769cd4d81ad2'
2022-08-13T14:34:32.8014374Z 
2022-08-13T14:34:32.8014617Z C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\shutil.py:627: PermissionError
2022-08-13T14:34:32.8014889Z 
2022-08-13T14:34:32.8015090Z During handling of the above exception, another exception occurred:
2022-08-13T14:34:32.8015323Z 
2022-08-13T14:34:32.8015781Z tmp_path = WindowsPath('C:/Users/runneradmin/AppData/Local/Temp/pytest-of-runneradmin/pytest-0/test_computes_and_stores_md5su0')
2022-08-13T14:34:32.8016102Z 
2022-08-13T14:34:32.8016202Z     @beartype
2022-08-13T14:34:32.8016509Z     def test_computes_and_stores_md5sum(tmp_path: Path):
2022-08-13T14:34:32.8016902Z         """Does ki add new hash to `.ki/hashes`?"""
2022-08-13T14:34:32.8017259Z         ORIGINAL: SampleCollection = get_test_collection("original")
2022-08-13T14:34:32.8017597Z         runner = CliRunner()
2022-08-13T14:34:32.8017940Z         with runner.isolated_filesystem(temp_dir=tmp_path):
2022-08-13T14:34:32.8018253Z             # Clone collection in cwd.
2022-08-13T14:34:32.8018566Z             clone(runner, ORIGINAL.col_file)
2022-08-13T14:34:32.8018825Z     
2022-08-13T14:34:32.8019169Z             # Check that hash is written.
2022-08-13T14:34:32.8019380Z             with open(
2022-08-13T14:34:32.8019792Z                 os.path.join(ORIGINAL.repodir, ".ki/hashes"), encoding="UTF-8"
2022-08-13T14:34:32.8020087Z             ) as hashes_file:
2022-08-13T14:34:32.8020314Z                 hashes = hashes_file.read()
2022-08-13T14:34:32.8020649Z                 assert f"a68250f8ee3dc8302534f908bcbafc6a  {ORIGINAL.filename}" in hashes
2022-08-13T14:34:32.8020940Z                 assert (
2022-08-13T14:34:32.8021223Z                     f"199216c39eeabe23a1da016a99ffd3e2  {ORIGINAL.filename}" not in hashes
2022-08-13T14:34:32.8021504Z                 )
2022-08-13T14:34:32.8021704Z     
2022-08-13T14:34:32.8021888Z             # Edit collection.
2022-08-13T14:34:32.8022277Z             shutil.copyfile(EDITED.path, ORIGINAL.col_file)
2022-08-13T14:34:32.8022512Z     
2022-08-13T14:34:32.8022734Z             logger.debug(f"CWD: {F.cwd()}")
2022-08-13T14:34:32.8022957Z     
2022-08-13T14:34:32.8023146Z             # Pull edited collection.
2022-08-13T14:34:32.8023404Z             os.chdir(ORIGINAL.repodir)
2022-08-13T14:34:32.8023643Z >           pull(runner)
2022-08-13T14:34:32.8023775Z 
2022-08-13T14:34:32.8023870Z tests\test_integration.py:184: 
2022-08-13T14:34:32.8024142Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2022-08-13T14:34:32.8024447Z <@beartype(tests.test_ki.pull) at 0x14f4b1ba9d0>:30: in pull
2022-08-13T14:34:32.8024698Z     ???
2022-08-13T14:34:32.8024889Z tests\test_ki.py:321: in pull
2022-08-13T14:34:32.8025217Z     res = runner.invoke(ki.ki, ["pull"], standalone_mode=False, catch_exceptions=False)
2022-08-13T14:34:32.8025887Z C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\click\testing.py:408: in invoke
2022-08-13T14:34:32.8026277Z     return_value = cli.main(args=args or (), prog_name=prog_name, **extra)
2022-08-13T14:34:32.8026791Z C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\click\core.py:1055: in main
2022-08-13T14:34:32.8027133Z     rv = self.invoke(ctx)
2022-08-13T14:34:32.8027575Z C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\click\core.py:1657: in invoke
2022-08-13T14:34:32.8027957Z     return _process_result(sub_ctx.command.invoke(sub_ctx))
2022-08-13T14:34:32.8028439Z C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\click\core.py:1404: in invoke
2022-08-13T14:34:32.8028819Z     return ctx.invoke(self.callback, **ctx.params)
2022-08-13T14:34:32.8029272Z C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\click\core.py:760: in invoke
2022-08-13T14:34:32.8029617Z     return __callback(*args, **kwargs)
2022-08-13T14:34:32.8029901Z <@beartype(ki.pull) at 0x14f4ae6faf0>:10: in pull
2022-08-13T14:34:32.8030557Z     ???
2022-08-13T14:34:32.8030927Z ki\__init__.py:1844: in pull
2022-08-13T14:34:32.8031169Z     _pull(kirepo, silent=False)
2022-08-13T14:34:32.8031423Z <@beartype(ki._pull) at 0x14f4ae6f9d0>:49: in _pull
2022-08-13T14:34:32.8031804Z     ???
2022-08-13T14:34:32.8032012Z ki\__init__.py:2136: in _pull
2022-08-13T14:34:32.8032316Z     last_push_root: NoFile = F.rmtree(F.working_dir(last_push_repo))
2022-08-13T14:34:32.8032647Z <@beartype(ki.functional.rmtree) at 0x14f4ae28700>:31: in rmtree
2022-08-13T14:34:32.8032903Z     ???
2022-08-13T14:34:32.8033117Z ki\functional.py:90: in rmtree
2022-08-13T14:34:32.8033339Z     rmtree2(str(target))
2022-08-13T14:34:32.8033632Z <@beartype(ki.functional.rmtree2) at 0x14f4adbf1f0>:30: in rmtree2
2022-08-13T14:34:32.8033894Z     ???
2022-08-13T14:34:32.8034084Z ki\functional.py:84: in rmtree2
2022-08-13T14:34:32.8034413Z     shutil.rmtree(path, ignore_errors=False, onerror=handle_remove_readonly)
2022-08-13T14:34:32.8034809Z C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\shutil.py:759: in rmtree
2022-08-13T14:34:32.8035113Z     return _rmtree_unsafe(path, onerror)
2022-08-13T14:34:32.8035456Z C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\shutil.py:624: in _rmtree_unsafe
2022-08-13T14:34:32.8035790Z     _rmtree_unsafe(fullname, onerror)
2022-08-13T14:34:32.8036272Z C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\shutil.py:624: in _rmtree_unsafe
2022-08-13T14:34:32.8036711Z     _rmtree_unsafe(fullname, onerror)
2022-08-13T14:34:32.8037054Z C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\shutil.py:624: in _rmtree_unsafe
2022-08-13T14:34:32.8037381Z     _rmtree_unsafe(fullname, onerror)
2022-08-13T14:34:32.8037699Z C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\shutil.py:629: in _rmtree_unsafe
2022-08-13T14:34:32.8038049Z     onerror(os.unlink, fullname, sys.exc_info())
2022-08-13T14:34:32.8038336Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2022-08-13T14:34:32.8038492Z 
2022-08-13T14:34:32.8038691Z __beartype_func = <function rmtree2.<locals>.handle_remove_readonly at 0x0000014F4AEBC430>
2022-08-13T14:34:32.8039071Z __beartype_raise_exception = <function raise_pep_call_exception at 0x0000014F4ABC1670>
2022-08-13T14:34:32.8039582Z __beartype_object_1439991030256 = <class 'collections.abc.Callable'>
2022-08-13T14:34:32.8039982Z __beartype_object_140725631987344 = <class 'NoneType'>
2022-08-13T14:34:32.8040757Z args = (<built-in function unlink>, 'C:\\Users\\runneradmin\\AppData\\Local\\Temp\\tmpqv92rnle\\ki-local-199216c39eeabe23a1da...81ad2', (<class 'PermissionError'>, PermissionError(13, 'Access is denied'), <traceback object at 0x0000014F4AEBA7C0>))
2022-08-13T14:34:32.8041323Z kwargs = {}, __beartype_args_len = 3
2022-08-13T14:34:32.8041836Z __beartype_pith_0 = (<class 'PermissionError'>, PermissionError(13, 'Access is denied'), <traceback object at 0x0000014F4AEBA7C0>)
2022-08-13T14:34:32.8042300Z __beartype_pith_1 = <class 'PermissionError'>
2022-08-13T14:34:32.8042571Z 
2022-08-13T14:34:32.8042627Z >   ???
2022-08-13T14:34:32.8042937Z E   NameError: name 'traceback' is not defined
2022-08-13T14:34:32.8043105Z 
2022-08-13T14:34:32.8043312Z <@beartype(ki.functional.rmtree2.handle_remove_readonly) at 0x14f4aebc430>:76: NameError
```

**Comment by leycec** [comment]
**Canadian Gods of the wilderness above,** I invoke thee! Save me from the horrifying consequences of my own bugs! `</ahem>`

What I *meant* to say is that I can all-too-sadly replicate this. Since your minimal length example embeds the clever landmine of recursively removing my entire filesystem, I lightly massaged everything explosive away as follows:

```python
import sys
from beartype import beartype
from beartype.typing import Any, Callable, Tuple, Type
from types import TracebackType

@beartype
def rmtree2(path: str) -> None:
    """On windows, rmtree fails for readonly dirs."""

    @beartype
    def handle_remove_readonly(
        func: Callable[..., Any],
        path: str,
        exc: Tuple[Type[OSError], OSError, TracebackType],
    ) -> None:
        pass

    try:
        raise OSError('Ugh!')
    except Exception as exception:
        handle_remove_readonly(
            lambda: 'Yamo!', path, (OSError, exception, sys.exc_info()[2]))

rmtree2('/tmp/bologo')
```

Things then blow up:

```
Traceback (most recent call last):
  File "/home/leycec/tmp/mopy.py", line 29, in rmtree2
    raise OSError('Ugh!')
OSError: Ugh!

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/leycec/tmp/mopy.py", line 34, in <module>
    rmtree2('/tmp/bologo')
  File "<@beartype(__main__.rmtree2) at 0x7f09bb472d40>", line 30, in rmtree2
  File "/home/leycec/tmp/mopy.py", line 31, in rmtree2
    handle_remove_readonly(
  File "<@beartype(__main__.rmtree2.handle_remove_readonly) at 0x7f09bb136b00>", line 76, in handle_remove_readonly
NameError: name 'traceback' is not defined
```

## I Can Explain

@beartype handles builtin types (e.g., `bool`, `str`) uniquely, because those types do *not* need to be explicitly passed to @beartype-decorated wrapper functions. Sadly, many types lie to @beartype by claiming that they're builtin despite not actually being builtin. It would seem that `types.TracebackType` is yet another such fake builtin:

```python
>>> from types import TracebackType
>>> TracebackType.__module__
'builtins'  # <-- curse you, Guido. curse you to all heck.
```

Thankfully, the fix is trivial. I'll have this patched up for everybody in a hot minute. :clock1:

**Comment by leycec** [comment]
**Resolved by d1f553c.** Sadly for us all, the next stable release of `beartype` is *still* several weeks away. Consider installing commit d1f553c instead via this one-liner as a temporary expedient:

```bash
$ python -m pip install git+https://github.com/beartype/beartype.git@d1f553c4fd5c8b392513ccd79c97a329f12ad626
```

If you bravely do so, you'll also get this super duper @beartype decorator that can **decorate classes** for free! Yes. It's all true. As seen live on late-night infomercials:

```python
from beartype import beartype

@beartype
class ThisNowWorksYo(object):
    def this_is_now_beartyped(
        self, oh: list[str], my: str, gods: tuple[bool, ...]) -> None:
        ...
```

:open_mouth:
